### PR TITLE
aspire: Support inputs and improve container.v0 support

### PR DESCRIPTION
**Pull request by ellismg**
This change adds supports for `inputs` which is a new (optional) property on resources in the Aspire manifest. These inputs represent values that deployment tools like `azd` need to provide "somehow".

For our initial support, we support string inputs which contain metadata on how to generate a default value, based on a minumum random length. Aspire uses this today to model passwords for cases like `AddMySqlContainer()` where a deployment tool needs to provide a password for the root user.

Consider the following AppHost manifest:

```json
{
  "resources": {
    "demoapp": {
      "type": "project.v0",
      "path": "../AspireContainers.DemoApp/AspireContainers.DemoApp.csproj",
      "env": {
        "ConnectionStrings__MySql": "{mysqlabstract.connectionString}"
      },
      "bindings": {
        "http": {
          "scheme": "http",
          "protocol": "tcp",
          "transport": "http",
        }
      }
    },
    "mysqlabstract": {
      "type": "container.v0",
      "image": "mysql:latest",
      "env": {
        "MYSQL_ROOT_PASSWORD": "{mysqlabstract.inputs.password}"
      },
      "bindings": {
        "tcp": {
          "scheme": "tcp",
          "protocol": "tcp",
          "transport": "tcp",
          "containerPort": 3306
        }
      },
      "connectionString": "Server={mysqlabstract.bindings.tcp.host};Port={mysqlabstract.bindings.tcp.port};User ID=root;Password={mysqlabstract.inputs.password}",
      "inputs": {
        "password": {
          "type": "string",
          "secret": true,
          "default": {
            "generate": {
              "minLength": 10
            }
          }
        }
      }
    }
  }
}
```

In this example, we spin up a MySql Container (with a randomly generated password at least 10 characters long). The container resource now has an expression with binding references in for its `connectionString` property, and other resources (like `demoapp`) can refer to this property to create connections to it.

To support this, we need to teach `azd` how to handle these inputs and where to store them as well as decide how to the flow into both the infrastructure and service deployments.

The strategy here for now is to model a new `Inputs` object, which holds all the values for the given inputs. The object contains a top level key for every resource with an input and the value is an object that maps inputs from that resource to their values, eg:

```json
{
  "mysqlabstract": {
     "password": "<value-from-deployment-tool>"
  }
}
```

For bicep deployments, you can mark an object or secure object parameter with `@metdata(azd: { type: 'inputs' } )` the value will be passed as this object.

For container app deployments, the new `Inputs` field is present on the root context object and you can read values from this map.

These are stored in the environment config (so they can vary per environment) under the "inputs" section.

Values which are marked as `secure` today are not encrypted at rest, but this is something we will improve before GA, by storing these values in KeyVault and then reference from there.

We do, however, mark the parameter that gets the value in the bicep deployment as an `@secure()` parameter, so the values like passwords are not retained in the deployment log.

As part of this change, some small improvements where made to the binding expression evaluator, to allow using `.host` in references to container apps (we know the value will be the same as the container app name, due to how networking works in ACA).

### CODE REVIEW

**Comment by ellismg** [comment]
This is the best I figure I can do for Preview 2.

There's still one small issue in the `infra synth` case - Once your run `infra synth` we will not correctly generate inputs during `azd provision` (if they are not already in the config) in this case, because we "lose track" of the tie between the infrastructure and the aspire manifest.  But `infra synth` is in alpha anyway, so I'm not super concerned.  As a workaround, one can manually edit the environment's `config.json` to add these inputs.

I think that long term in `infra synth` we want to write the metadata for the inputs somewhere so we can use it when loading the infrastructure later.

For Bicep it makes sense to store the metadata on the `inputs` property itself (or move to a model where each input gets its own bicep property, which might be nicer).

For the container-app.yaml files, I suspect we'll just want to generate some file like `.azd.metdata.json` or something SxS with the container-app.yaml file and that can hold whatever info we need.

**Comment by ellismg** [comment]
The change in #3130 will impact the baselines here, so we will need to re-generate them after that change lands.

File: cli/azd/resources/apphost/templates/resources.bicept
Comment (weikanglim) [inline cli/azd/resources/apphost/templates/resources.bicept:161]: This may be a bug fix for something else, but just confirming the intent of the change.

**Comment by ellismg** [comment]
Rebased to include the fix from #3130 so I could update my baseline.  Will remove the commits when that change goes in.  It does not clutter the diff too much, but you can also just look at the last commit which has all the net new changes.

File: cli/azd/pkg/apphost/generate_types.go
Comment (ellismg) [inline cli/azd/pkg/apphost/generate_types.go:24]: Without this, the template evaluation logic in my tests was failing because bicep template looks at the Dapr property of the context object, but it was not present.  Looking at the change that introduced this, it wasn't clear there were things to do on the Bicep side of things, but perhaps, @philliphoff understands more.  Would there be a case, @philliphoff where a `container.v0` resource (i.e. an existing container image) would have dapr stuff added to it?  I can image that would be the case - if so, I think we need to make some more changes to ensure that this value gets set correctly.

File: cli/azd/cmd/provision.go
Comment (weikanglim) [reply cli/azd/cmd/provision.go:173]: I think we only assume `inputs` have these sections attached for now, and simply panic if unsupported?

**Review by weikanglim — COMMENTED** [review]
Finished reviewing. I felt like I learned a lot reviewing this change! Will perform some testing before approving...

File: cli/azd/resources/apphost/templates/resources.bicept
Comment (ellismg) [reply cli/azd/resources/apphost/templates/resources.bicept:161]: Indeed - this was fixing a pre-existing issue that I discovered while doing some end to end testing myself.  The problem was before we would always set `allowInsecure` (which tells ACA if the Envoy proxy that fronts each host should allow HTTP on port 80 (`true`) or if it should instead preform an HTTP-HTTPS redirection (`false`).

For containers that expose just raw TCP ports, there is no load balancer and ACA fails the request if the `allowInsecure` value is set.  So this change ensures we exclude it on these types of ingress configurations.

File: cli/azd/cmd/provision.go
Comment (ellismg) [reply cli/azd/cmd/provision.go:173]: Yeah - corners cut here due to the enforcement here:

https://github.com/Azure/azure-dev/pull/3131/files#diff-b711e594f7ac2f56ef4386a9fad07d756710f296fc0a2496c3148f6ec0c1b821R283-R293

Which validates that ever input can be generated today and has a min length.  There's a TODO around that code, but I'll add some notes here as well as to why this dereference is safe today and how it will need to be improved when we relax this requirement.

**Review by weikanglim — APPROVED** [review]
Finished testing and didn't catch anything obvious. This was a humbling experience for me, as I learned about how to accomplish a lot with so little. I think it also speaks to the well-thought-out, slim and flexible design of the components built to support the Aspire workload.

Thank you for getting this change out! LGTM!

File: cli/azd/pkg/password/generator.go
Comment (ellismg) [reply cli/azd/pkg/password/generator.go:27]: Total fat finger on my part here.  I wanted <= and just typed the wrong thing.  Good catch, will fix!

File: cli/azd/cmd/provision.go
Comment (ellismg) [reply cli/azd/cmd/provision.go:178]: Added some load bearing TODO's...

File: cli/azd/pkg/apphost/generate.go
Comment (davidfowl) [inline cli/azd/pkg/apphost/generate.go:349]: Do you want this version to break when using it on preview1 projects?

Comment (ellismg) [reply cli/azd/pkg/apphost/generate.go:349]: I did not see a reason to have stuff break - we now handle the connectionString property in a more uniform way.  I suspect for preview3 we'll remove the code for these old connection type resources you no longer emit, but I figured for now we might as well support the old version as well when there was no cost.

Comment (davidfowl) [reply cli/azd/pkg/apphost/generate.go:349]: Ok great! I agree with you.

**Pull request merged**
